### DISCUSSION TITLE
GitHub provider : "No pre-authenticated principal found in request"

### DISCUSSION THREAD

**Discussion by rwithers** [IridiumIdentity/iridium#85] [Q&A]
This post is a question about going password less.  I'm initializing with the github provider choosing the login flow (the system asks me to authenticate with github and instead of taking me into iridium I'm getting getting redirected back to the login page, and I don't know why.  Below I've listed in a pretty detailed fashion the steps I'm going through.  I'm happy to elaborate where necessary. 

I've reloaded with the most recent changes figuring I must have done something wrong (this is after the password removal changes).  What I'm seeing in the logs is: "no pre-authenticated principal found in request."  I have also included a snippet of logs from the container runtime. 

I can't figure out the forwarding / redirecting thing.  It dumps me right back to the register/login page when I try to login with github.  

I wanted to list out the path I took, to see if you can see anything that I'm clearly missing.  I'm backing all the way up to the point of building the image with the newly built jar from source.  But I'm not including those steps which would be the mvn commands. 

1. cd to projects/iridium/tools/images/
2. cp iridium-core-server-0.1.4.jar (previously built from source) 
3. execute docker build: docker build -f Dockerfile.core . (from directory in step 1) 
4. docker tag <hash> iridiumidentiy/iridium-core-server:latest
5. Then I verified the docker compose file and ran the following command: docker compose -f tools/schedulers/compose/local-iridium-compose.yml up
6. Steps 3, 4, and 5 enabled me to run the latest build 0.1.4
7. Then I ran the initialization step 
/bin/iridium init \
--host=localhost \
--port=3306 \
--user=root \
--password \
--allow-github=true
8. At this point I went to my github instance which I have configured exactly like the "readthedocs" suggest: 

Application Name: IridiumTestInstance
Homepage URL: http://localhost:4200/
Application Description: Iridium Testing
Application Callback URL: http://localhost:4200/dashboard

I copied the client id and a newly generated secret when prompted by the init command above (step 7) I'm not including those there for obvious reasons. 

I don't have any local changes, this is just a pull from main authenticating using the github provider.  I figured before I can confidently submit code I need to be able to duplicate the functionality utilizing the newest codebase.  

Did I miss anything? 

Thanks, 

Ryan 

iridium  | 2023-06-25T15:43:58.907Z DEBUG 1 --- [io-8381-exec-10] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to software.iridium.api.controller.TenantController#getTenantSummaries(HttpServletRequest)
iridium  | 2023-06-25T15:43:58.907Z DEBUG 1 --- [nio-8381-exec-9] m.m.a.RequestResponseBodyMethodProcessor : Using 'application/vnd.iridium.id.identity-response.1+json', given [application/vnd.iridium.id.identity-response.1+json] and supported [application/json, application/*+json]
iridium  | 2023-06-25T15:43:58.907Z DEBUG 1 --- [nio-8381-exec-9] m.m.a.RequestResponseBodyMethodProcessor : Writing [software.iridium.api.base.domain.ApiResponse@8b0943b]
iridium  | 2023-06-25T15:43:58.908Z DEBUG 1 --- [io-8381-exec-10] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to software.iridium.api.controller.TenantController#getTenantSummaries(HttpServletRequest)
iridium  | 2023-06-25T15:43:58.908Z DEBUG 1 --- [io-8381-exec-10] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to software.iridium.api.controller.TenantController#getTenantSummaries(HttpServletRequest)
iridium  | 2023-06-25T15:43:58.908Z DEBUG 1 --- [io-8381-exec-10] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to software.iridium.api.controller.TenantController#getTenantSummaries(HttpServletRequest)
iridium  | 2023-06-25T15:43:58.909Z DEBUG 1 --- [io-8381-exec-10] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to software.iridium.api.controller.TenantController#getTenantSummaries(HttpServletRequest)
iridium  | 2023-06-25T15:43:58.909Z DEBUG 1 --- [nio-8381-exec-9] .m.m.a.ExceptionHandlerExceptionResolver : Resolved [software.iridium.api.base.error.NotAuthorizedException: NOT AUTHORIZED]
iridium  | 2023-06-25T15:43:58.910Z DEBUG 1 --- [nio-8381-exec-9] o.s.web.servlet.DispatcherServlet        : Completed 401 UNAUTHORIZED
iridium  | 2023-06-25T15:43:58.910Z DEBUG 1 --- [io-8381-exec-10] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to software.iridium.api.controller.TenantController#getTenantSummaries(HttpServletRequest)
iridium  | 2023-06-25T15:43:58.911Z  INFO 1 --- [nio-8381-exec-9] o.s.web.filter.OncePerRequestFilter      : After request - URI: /identities, Method: GET, Payload: No Payload, Time to process: 27ms
iridium  | 2023-06-25T15:43:58.911Z  INFO 1 --- [io-8381-exec-10] o.s.web.filter.OncePerRequestFilter      : After request - URI: /tenants, Method: GET, Payload: No Payload, Time to process: 10ms
iridium  | 2023-06-25T15:43:58.913Z DEBUG 1 --- [io-8381-exec-10] s.iridium.api.TokenAuthenticationFilter  : Authenticating null
iridium  | 2023-06-25T15:43:58.913Z DEBUG 1 --- [io-8381-exec-10] s.iridium.api.TokenAuthenticationFilter  : No pre-authenticated principal found in request

**Comment by joshfischer1108** [comment]
I have a few questions:

* Are you building the CLI from source or downloading it from the release artifacts?

In reference to... 
```
Application Name: IridiumTestInstance
Homepage URL: http://localhost:4200/
Application Description: Iridium Testing
Application Callback URL: http://localhost:4200/dashboard
```

Are these the settings in  Github or are they set in Iridium?   I suspect you might need to make the callback URI `http://localhost:4200/callback`.  If you are trying to get the management UI working. 

Thanks for reporting this, I see some holes in the docs that need to be filled.

**Comment by rwithers** [comment]
Okay the dashboard url is coming from the docs.



On Sun, Jun 25, 2023 at 8:42 PM Josh Fischer ***@***.***>
wrote:

> I have a few questions:
>
>    - Are you building the CLI from source or downloading it from the
>    release artifacts?
>
> In reference to...
>
> Application Name: IridiumTestInstance
> Homepage URL: http://localhost:4200/
> Application Description: Iridium Testing
> Application Callback URL: http://localhost:4200/dashboard
>
> Are these the settings in Github or are they set in Iridium? I suspect you
> might need to make the callback URI http://localhost:4200/callback. If
> you are trying to get the management UI working.
>
> Thanks for reporting this, I see some holes in the docs that need to be
> filled.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/IridiumIdentity/discussions/85#discussioncomment-6277029>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24WDMSBM6XKENFPISS27TXNDSHTANCNFSM6AAAAAAZTJFHRQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Ryan Withers
Senior Software Developer / Analyst

http://www.linkedin.com/in/ryanwithers
Work: (314) 799-5817

**Comment by rwithers** [comment]
The cli is being built from source and the bin is being unpacked and run
relatively from the target directory.



On Sun, Jun 25, 2023 at 10:04 PM Ryan Withers ***@***.***>
wrote:

> Okay the dashboard url is coming from the docs.
>
>
>
> On Sun, Jun 25, 2023 at 8:42 PM Josh Fischer ***@***.***>
> wrote:
>
>> I have a few questions:
>>
>>    - Are you building the CLI from source or downloading it from the
>>    release artifacts?
>>
>> In reference to...
>>
>> Application Name: IridiumTestInstance
>> Homepage URL: http://localhost:4200/
>> Application Description: Iridium Testing
>> Application Callback URL: http://localhost:4200/dashboard
>>
>> Are these the settings in Github or are they set in Iridium? I suspect
>> you might need to make the callback URI http://localhost:4200/callback.
>> If you are trying to get the management UI working.
>>
>> Thanks for reporting this, I see some holes in the docs that need to be
>> filled.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/orgs/IridiumIdentity/discussions/85#discussioncomment-6277029>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AD24WDMSBM6XKENFPISS27TXNDSHTANCNFSM6AAAAAAZTJFHRQ>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
> --
> Ryan Withers
> Senior Software Developer / Analyst
>
> http://www.linkedin.com/in/ryanwithers
> Work: (314) 799-5817
>
-- 
Ryan Withers
Senior Software Developer / Analyst

http://www.linkedin.com/in/ryanwithers
Work: (314) 799-5817

**Comment by joshfischer1108** [comment]
After looking at the logs I see 
```
software.iridium.api.controller.TenantController#getTenantSummaries(HttpServletRequest)
iridium | 2023-06-25T15:43:58.909Z DEBUG 1 --- [nio-8381-exec-9] .m.m.a.ExceptionHandlerExceptionResolver : Resolved [software.iridium.api.base.error.NotAuthorizedException: NOT AUTHORIZED]
```

Is it possible your `iridium-token` being passed was valid?  This can happen when the server is stopped and restarted or the database is re-initted.

To verify, you can check the value stored in your browser vs the value stored in the database in the `access_tokens` table.

e.q.
```
select * from access_tokens;
```

**Comment by rwithers** [comment]
I will go through the steps again and select as you suggest both before and
after registration attempt.

How long is it valid, is the process for refreshing just logging in again?




On Mon, Jun 26, 2023 at 6:36 AM Josh Fischer ***@***.***>
wrote:

> After looking at the logs I see
>
> software.iridium.api.controller.TenantController#getTenantSummaries(HttpServletRequest)
> iridium | 2023-06-25T15:43:58.909Z DEBUG 1 --- [nio-8381-exec-9] .m.m.a.ExceptionHandlerExceptionResolver : Resolved [software.iridium.api.base.error.NotAuthorizedException: NOT AUTHORIZED]
>
> Is it possible your iridium-token being passed was valid? This can happen
> when the server is stopped and restarted or the database is re-initted.
>
> To verify, you can check the value stored in your browser vs the value
> stored in the database in the access_tokens table.
>
> e.q.
>
> select * from access_tokens;
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/IridiumIdentity/discussions/85#discussioncomment-6281244>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24WDNUVJKX4XJTHYEQNTLXNFX5JANCNFSM6AAAAAAZTJFHRQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Ryan Withers
Senior Software Developer / Analyst

http://www.linkedin.com/in/ryanwithers
Work: (314) 799-5817

**Reply by rwithers** [reply]
This is the latest and the actual error that preceded the message about "No pre-authenticated principal"

iridium  | 2023-06-27T01:30:11.914Z  INFO 1 --- [nio-8381-exec-9] o.s.web.filter.OncePerRequestFilter      : After request - URI: /oauth/token?grant_type=authorization_code&code=a67385587395dd8c9822&redirect_uri=http://localhost:4200/callback&client_id=, Method: POST, Payload: No Payload, Time to process: 534ms
iridium  | 2023-06-27T01:30:11.915Z ERROR 1 --- [nio-8381-exec-9] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: software.iridium.api.base.error.ClientCallException: Failure posting to provider] with root cause
iridium  |
iridium  | org.springframework.web.client.HttpClientErrorException$Unauthorized: 401 Unauthorized: "{"message":"Bad credentials","documentation_url":"https://docs.github.com/rest"}"
iridium  | 	at org.springframework.web.client.HttpClientErrorException.create(HttpClientErrorException.java:106) ~[spring-web-6.0.8.jar!/:6.0.8]
iridium  | 	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:183) ~[spring-web-6.0.8.jar!/:6.0.8]
iridium  | 	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:137) ~[spring-web-6.0.8.jar!/:6.0.8]
iridium  | 	at org.springframework.web.client.ResponseErrorHandler.handleError(ResponseErrorHandler.java:63) ~[spring-web-6.0.8.jar!/:6.0.8]
iridium  | 	at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:915) ~[spring-web-6.0.8.jar!/:6.0.8]
iridium  | 	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:864) ~[spring-web-6.0.8.jar!/:6.0.8]
iridium  | 	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:804) ~[spring-web-6.0.8.jar!/:6.0.8]
iridium  | 	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:695) ~[spring-web-6.0.8.jar!/:6.0.8]
iridium  | 	at software.iridium.api.authentication.client.ProviderProfileRequestor.requestGithubProfile(ProviderProfileRequestor.java:50) ~[iridium-java-client-0.1.4.jar!/:0.1.4]
